crates/csharp/src/lib.rs

```diff
@@ -2018,23 +2089,89 @@ impl Bindgen for FunctionBindgen<'_, '_> {
                 self.gen.gen.needs_interop_string = true;
             }
 
-            Instruction::StringLift { .. } => match self.gen.direction {
-                Direction::Import => results.push(format!("ReturnArea.GetUTF8String(ptr)")),
-                Direction::Export => {
-                    let address = &operands[0];
-                    let length = &operands[1];
+            Instruction::StringLift { .. } => results.push(format!(
+                "Encoding.UTF8.GetString((byte*){}, {})",
+                operands[0], operands[1]
+            )),
 
-                    results.push(format!("returnArea.GetUTF8String({address}, {length})"));
-                }
-            },
+            Instruction::ListLower { element, realloc } => {
+                let Block {
+                    body,
+                    results: block_results,
+                    element: block_element,
+                    base,
+                } = self.blocks.pop().unwrap();
+                assert!(block_results.is_empty());
+
+                let op = &operands[0];
+                let size = self.gen.gen.sizes.size(element);
+                let align = self.gen.gen.sizes.align(element);
+                let ty = self.gen.type_name(element);
+                let index = self.locals.tmp("index");
+                let list: String = self.locals.tmp("list");
+
+                let buffer: String = self.locals.tmp("buffer");
+
+                uwrite!(
+                    self.src,
+                    "
+                    void* {buffer} = stackalloc byte[{size} + {align} * ({op}).Count];
```

Comment:
I think we need to allocate this buffer from the heap.  When this method returns the buffer is going to be dangling and in danger or being overwritten.

Reply:
Since this is on the import direction I thought the stack was OK? We are doing this for the return area as well

Reply:
Have I marked the wrong rust code.  I meant to mark what generates this export:
```cs
        unsafe public static int wasmExportListResult3() {
            var ret = TestImpl.ListResult3();
            
            var ptr = InteropReturnArea.returnArea.AddressOfReturnArea();
            
            void* buffer = stackalloc byte[8 + 4 * (ret).Count];
            var list = ret;
            
            for (int index = 0; index < list.Count; ++index) {
                string element = list[index];
                int basePtr = (int)buffer + (index * 8);
                
                var result = element;
                IntPtr interopString = InteropString.FromString(result, out int lengthresult);
                BitConverter.TryWriteBytes(new Span<byte>((void*)(basePtr + 4), 4), unchecked((int)lengthresult));
                BitConverter.TryWriteBytes(new Span<byte>((void*)(basePtr + 0), 4), unchecked((int)interopString.ToInt32()));
                
            }
            BitConverter.TryWriteBytes(new Span<byte>((void*)(ptr + 4), 4), unchecked((int)list.Count));
            BitConverter.TryWriteBytes(new Span<byte>((void*)(ptr + 0), 4), unchecked((int)(int)buffer));
            return ptr;
            
        }
```

Reply:
In this case we are using [`BitConverter.TryWriteBytes`](https://learn.microsoft.com/en-us/dotnet/api/system.bitconverter.trywritebytes?view=net-8.0) which will write the value to the destination provided at ptr, so we are writing the value onto the returnArea which is on the heap so we should be ok.

Reply:
But aren't we writing the address of `buffer` which is a stack location into the return area?

Reply:
I've updated this, I think I mis-understood the return pointer scenario and this scenario which aren't the same.